### PR TITLE
Add @page directive

### PR DESCRIPTION
**Pull request by rynowak**
Adds the @page directive and support for specifying routes in components
at compile time.

For now the route is required and must begin with a leading /.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Blazor.Razor.Extensions/PageDirective.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+{
+    public class PageDirective
+    {
+        public static readonly DirectiveDescriptor Directive = DirectiveDescriptor.CreateDirective(
+            "page",
+            DirectiveKind.SingleLine,
+            builder =>
+            {
+                builder.AddStringToken(Resources.PageDirective_RouteToken_Name, Resources.PageDirective_RouteToken_Description);
+                builder.Usage = DirectiveUsage.FileScopedSinglyOccurring;
```
Comment (SteveSandersonMS) [inline src/Microsoft.AspNetCore.Blazor.Razor.Extensions/PageDirective.cs]: Could we have `FileScopedMultipleOccurring`? There will be cases where people want to define more than one URL pattern that maps to the same component.

File: src/Microsoft.AspNetCore.Blazor.Razor.Extensions/PageDirectivePass.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                var directive = directives[i];
+                if (directive.Node.IsImported())
+                {
+                    directive.Node.Diagnostics.Add(BlazorDiagnosticFactory.CreatePageDirective_CannotBeImported(directive.Node.Source.Value));
+                }
+            }
+
+            // Only process the first directive occurrence. The parser will have already flagged multiple
+            // usages as errors.
+            var pageDirective = (DirectiveIntermediateNode)directives[0].Node;
```
Comment (SteveSandersonMS) [inline src/Microsoft.AspNetCore.Blazor.Razor.Extensions/PageDirectivePass.cs]: As previous comment, would be good if we allowed multiple.

File: src/Microsoft.AspNetCore.Blazor/Components/RouteAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Blazor.Components
+{
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = false, Inherited = true)]
```
Comment (SteveSandersonMS) [inline src/Microsoft.AspNetCore.Blazor/Components/RouteAttribute.cs]: Also here

File: src/Microsoft.AspNetCore.Blazor.Razor.Extensions/PageDirective.cs (hunk 1 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Blazor.Razor.Extensions/PageDirective.cs]: Sure

**Pull request merged**